### PR TITLE
Helper scripts for profiling

### PR DESCRIPTION
**Pull request by adobrzyniewicz-habana**
Create helper scripts for profiling

To capture a single forward to profiler file run:
`python3 hpu-utils/profile_forward.py --model-path "/mnt/weka/data/pytorch/llama2/Llama-2-7b-chat-hf/"` 

Sample run using custom flags:
`python3 hpu-utils/profile_forward.py --model-path "/mnt/weka/data/pytorch/llama2/Llama-2-70b-chat-hf/" --num-cards 8 --output-path "./llama_70b_output.json.gz" --batch-size 256`

All flags:
--model-path
--num-cards
--phase
--data-type
--output-path
--block-size
--batch-size
--seq-len
--steps

### CODE REVIEW

File: run_vllm_forward.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        record_shapes=False,
+        with_stack=True)
+    return profiler
+
+def kill_process():
+    """Kills python3 main process manually"""
+    print("Killing process manually")
+    import psutil
+    for proc in psutil.process_iter():
+        if proc.name() == "python3":
```
Comment (madamczykhabana) [inline run_vllm_forward.py:31]: That's a bit brutal... Could we just kill 'ray' instances? Besides this can lead to a race condition when this script gets killed before ray.

File: run_vllm_forward.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        seq = SequenceGroupMetadata(
+            request_id=str(group_id),
+            is_prompt=(output_len == 0),
+            seq_data={group_id: seq_data},
+            sampling_params=sampling_params,
+            block_tables=block_tables,
+        )
+        seqs.append(seq)
+
+    model_request = ExecuteModelRequest(seq_group_metadata_list=seqs)
```
Comment (madamczykhabana) [inline run_vllm_forward.py:64]: Hmm, we already have functions that create dummy requests in habana_runner. Perhaps we could call them here directly to create the requests?

File: run_vllm_forward.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    print("Finished running llm")
+    if args.num_cards > 1:
+        kill_process()
+
+parser = argparse.ArgumentParser("vLLM arguments parser")
+
+parser.add_argument("--model", help="Path to the model that will be used", type=str)
+parser.add_argument("--num-cards", help="Number of cards that will be used by model", type=int)
+parser.add_argument("--phase", help="Phase", type=str)
+parser.add_argument("--eager", help="Run in eager mode", type=bool)
```
Comment (madamczykhabana) [inline run_vllm_forward.py:89]: We can ignore this parameter. We want to profile a single step, and HPU graphs won't be enabled globally without a full warmup anyway.

File: run_vllm_forward.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+if args.data_type == "bf16":
+    model_dtype = torch.bfloat16
+else:
+    print("Invalid data type, using bf16")
+    model_dtype = torch.bfloat16
+
+if args.phase == "prompt":
+    is_prompt = True
+else:
+    is_prompt = False
```
Comment (madamczykhabana) [inline run_vllm_forward.py:108]: `is_prompt = arg.phase == "prompt"`

File: run_vllm_forward.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+parser.add_argument("--model", help="Path to the model that will be used", type=str)
+parser.add_argument("--num-cards", help="Number of cards that will be used by model", type=int)
+parser.add_argument("--phase", help="Phase", type=str)
+parser.add_argument("--eager", help="Run in eager mode", type=bool)
+parser.add_argument("--data-type", help="Type of data that will be used", type=str)
+parser.add_argument("--block-size", help="Block size", type=int)
+parser.add_argument("--batch-size", help="Batch size", type=int)
+parser.add_argument("--seq-len", help="Sequence length", type=int)
+parser.add_argument("--steps", help="Number of steps", type=int)
```
Comment (madamczykhabana) [inline run_vllm_forward.py:94]: I'd add a default value for this. '3' seams reasonable

File: run_vllm_forward.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    profiler.stop()
+    print("Finished running llm")
+    if args.num_cards > 1:
+        kill_process()
+
+parser = argparse.ArgumentParser("vLLM arguments parser")
+
+parser.add_argument("--model", help="Path to the model that will be used", type=str)
+parser.add_argument("--num-cards", help="Number of cards that will be used by model", type=int)
+parser.add_argument("--phase", help="Phase", type=str)
```
Comment (madamczykhabana) [inline run_vllm_forward.py:88]: `choices=["prompt", "decode"]`

File: run_vllm_forward.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+parser.add_argument("--steps", help="Number of steps", type=int)
+args = parser.parse_args()
+
+print(args)
+
+if args.data_type == "bf16":
+    model_dtype = torch.bfloat16
+else:
+    print("Invalid data type, using bf16")
+    model_dtype = torch.bfloat16
```
Comment (madamczykhabana) [inline run_vllm_forward.py:103]: I'd do the checks by limiting possible choices. For now we can go with `choices=["bf16"]` . Once fp8 is enabled we'll add it.

File: run_vllm_forward.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+if args.phase == "prompt":
+    is_prompt = True
+else:
+    is_prompt = False
+
+run_vllm()
```
Comment (madamczykhabana) [inline run_vllm_forward.py:110]: We can pass args object directly to run_vllm. No need to use globals

File: run_vllm_forward.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+--data-type ${DATA_TYPE} \
+--block-size ${BLOCK_SIZE} \
+--batch-size ${BATCH_SIZE} \
+--seq-len ${SEQ_LEN} \
+--steps ${STEPS} 
+
+if [ $? ];
+then
+	PROFILER_FILE=$(ls -t | head -1)
+	mv $PROFILER_FILE $OUTPUT_PATH
```
Comment (madamczykhabana) [inline run_vllm_forward.sh:45]: Moving *.json.gz to the destination could be handled by run_forward.py. This way we wouldn't need this script at all.

File: run_vllm_forward.sh [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+--eager ${EAGER} \
+--data-type ${DATA_TYPE} \
+--block-size ${BLOCK_SIZE} \
+--batch-size ${BATCH_SIZE} \
+--seq-len ${SEQ_LEN} \
+--steps ${STEPS} 
+
+if [ $? ];
+then
+	PROFILER_FILE=$(ls -t | head -1)
```
Comment (madamczykhabana) [inline run_vllm_forward.sh:44]: We can't rely on timing information. We should rather clear all *.trace.json.gz files first and then copy the only one after capturing.

File: run_vllm_forward.py
Comment (madamczykhabana) [inline run_vllm_forward.py:1]: perhaps we could call it 'profile_forward.py' ?

Comment (madamczykhabana) [inline run_vllm_forward.py:1]: It would be nice to move it to a HPU-specific subdirectory. Something like 'hpu-utils'?

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

File: run_vllm_forward.sh [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+STEPS=${STEPS:-5}
+OUTPUT_PATH=${OUTPUT_PATH:-output.json.gz}
+# ------------------
+
+case $MODEL in
+	"llama-7b")
+	MODEL_PATH="/mnt/weka/data/pytorch/llama2/Llama-2-7b-chat-hf/"
+	;;
+	"llama-70b")
+	MODEL_PATH="/mnt/weka/data/pytorch/llama2/Llama-2-70b-chat-hf/"
```
Comment (madamczykhabana) [inline run_vllm_forward.sh:24]: We can't assume that all users will have the weights in the same location. This should be passed as a parameter

File: run_vllm_forward.py (hunk 1 above)
Comment (kzawora-intel) [reply run_vllm_forward.py:31]: what about wrapping the job in separate process that we spawn and forcibly killing that process afterwards?

Comment (kzawora-intel) [reply run_vllm_forward.py:31]: something like this: https://github.com/vllm-project/vllm/issues/1908#issuecomment-2141297771

Comment (adobrzyniewicz-habana) [reply run_vllm_forward.py]: Tried solution from this issue, works better :)

File: run_vllm_forward.sh (hunk 9 above)
Comment (adobrzyniewicz-habana) [reply run_vllm_forward.sh]: Moved this to .py file

File: run_vllm_forward.sh (hunk 10 above)
Comment (adobrzyniewicz-habana) [reply run_vllm_forward.sh]: After changes I'm grabbing latest *.json.gz file

File: run_vllm_forward.py (hunk 2 above)
Comment (adobrzyniewicz-habana) [reply run_vllm_forward.py]: Decided to skip that for now

File: hpu-utils/profile_forward.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    print("Starting steps")
+    for _ in range(args.steps):
+        run_forward(llm, is_prompt, args.block_size, args.batch_size, args.seq_len)
+        profiler.step()
+    profiler.stop()
+    print("Finished running llm")
+
+parser = argparse.ArgumentParser("vLLM arguments parser")
+
+parser.add_argument("--model-path", help="Path to the model that will be used", type=str, default="/mnt/weka/data/pytorch/llama2/Llama-2-7b-chat-hf/")
```
Comment (madamczykhabana) [inline hpu-utils/profile_forward.py:86]: Let's remove that default as we shouldn't assume the user would have access to this path.

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

File: hpu-utils/profile_forward.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
+###############################################################################
+# Copyright (C) 2024 Habana Labs, Ltd. an Intel Company
+###############################################################################
+
+import argparse
+import torch
+import os
```
Comment (madamczykhabana) [inline hpu-utils/profile_forward.py:7]: As a convenience you can add: `os.environ['VLLM_SKIP_WARMUP']='true'` somewhere at the top before loading vllm.
This way the users don't need to remeber to do that manually.

File: hpu-utils/profile_forward.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    profiler = setup_profiler(args.steps)
+    profiler.start()
+    print("Starting steps")
+    for _ in range(args.steps):
+        run_forward(llm, is_prompt, args.block_size, args.batch_size, args.seq_len)
+        profiler.step()
+    profiler.stop()
+    print("Finished running llm")
+
+parser = argparse.ArgumentParser("vLLM arguments parser")
```
Comment (madamczykhabana) [inline hpu-utils/profile_forward.py:84]: You can add: `formatter_class=argparse.ArgumentDefaultsHelpFormatter`. This will print-out the defaults.
